pkg/utils/concurrency.go

```diff
@@ -0,0 +1,145 @@
+package utils
+
+import "sync"
+
+// This function is intended to be used in the situation where one has a slice of channels that will each eventually publish a single vue and then be closed,
+// 		and one needs to emit a single slice value on another channel that collapses the results down
+// Specifically, this is the most basic variant without error handling
+func SliceOfChannelsResultSynchronizer[IR any, FR any](individualResultChannels []<-chan IR,
+	finalResultChannel chan<- []FR, mappingFn func(IR) FR) {
+
+	var wg sync.WaitGroup
+	wg.Add(len(individualResultChannels))
+	resultMutex := sync.Mutex{}
+	result := []FR{}
+	for _, irc := range individualResultChannels {
+		go func(c <-chan IR) {
+			defer wg.Done()
+			ir := <-c
+			resultMutex.Lock()
+			result = append(result, mappingFn(ir))
+			resultMutex.Unlock()
+		}(irc)
+	}
+	wg.Wait()
+	finalResultChannel <- result

```

Comment:
@richiejp Thank you for the comments! I'm using this opportunity to learn about golang's channels and concurrency primitives myself,  but notably, these *Synchronizer functions were from an earlier revision... and I forgot to ever go back and remove them from the utils class after I removed them from the core. 

Rather than blocking at the level of `core/backend/llm.go` I've switched to a style of returning multiple channels - the caller can determine if it's best to block on the final result (non-streaming) or various subsets of partial results, like per-token or per-completion.